sdk/identity/azure-identity/azure/identity/_internal/aad_client_base.py

```diff
@@ -96,17 +89,43 @@ def _process_response(self, response, scopes, now):
             )
         return AccessToken(response_copy["access_token"], expires_on)
 
-    @abc.abstractmethod
-    def _get_client_session(self, **kwargs):
-        pass
-
-    @abc.abstractmethod
-    def _obtain_token(self, scopes, fn, **kwargs):
-        # type: (Iterable[str], Callable, **Any) -> AccessToken
-        pass
+    def _get_auth_code_request(self, scopes, code, redirect_uri, client_secret=None):
+        # type: (str, str, Sequence[str], Optional[str]) -> HttpRequest
+
+        data = {
+            "client_id": self._client_id,
+            "code": code,
+            "grant_type": "authorization_code",
+            "redirect_uri": redirect_uri,
+            "scope": " ".join(scopes),
+        }
+        if client_secret:
+            data["client_secret"] = client_secret
+
+        request = HttpRequest(
+            "POST", self._token_endpoint, headers={"Content-Type": "application/x-www-form-urlencoded"}
+        )
+        request.set_formdata_body(data)
+        return request
+
+    def _get_refresh_token_request(self, scopes, refresh_token):
+        # type: (str, Sequence[str]) -> HttpRequest
+
+        data = {
```

Comment:
What if refresh_token is None?

Reply:
If it serializes, you get an error from AAD.

Reply:
Then it will be an http error rather than a value error, right?

Reply:
Yes, assuming something doesn't raise before the request goes onto the wire.

Reply:
Is it possible empty refresh_token gets something back (the un-deterministic issue)?

Reply:
If it gets anything other than an error, that's an AAD bug.